### PR TITLE
Adding file monitoring hooks

### PR DESCRIPTION
**Pull request by corey-nm**
Note: Not integrated into server yet.

Main hook is `start_file_watcher` for server to call into to kick off a watcher process. Everything else is just helpers for that.

The file watcher at a high level does the following:
1. Check every X seconds if the file has been modified
2. If so, check if any endpoints were added/removed/modified
3. If endpoints have changed, send requests to server to update it live
4. Save the old config as `<path>.v<version>`

test plan: unit tests

### CODE REVIEW

File: src/deepsparse/server/config.py [hunk 1]
```diff
@@ -153,6 +153,28 @@ class ServerConfig(BaseModel):
     )
 
 
+def _endpoint_diff(
+    old: ServerConfig, new: ServerConfig
+) -> Tuple[List[EndpointConfig], List[EndpointConfig]]:
+    """ """
```
Comment (rahul-tuli) [inline src/deepsparse/server/config.py:159]: nit: remove empty docstring

File: src/deepsparse/server/config.py [hunk 2]
```diff
@@ -153,6 +153,28 @@ class ServerConfig(BaseModel):
...
+    old_routes = {e.route: e for e in old.endpoints if e.route is not None}
```
Comment (rahul-tuli) [inline src/deepsparse/server/config.py:160]: nit: rename `e`?

File: src/deepsparse/server/monitoring.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        ),
+        daemon=True,
+    )
+    proc.start()
+    return proc
+
+
+def _watch_file(config_path: str, endpoints_url: str, check_interval_s: float):
+    for _ in _diff_generator(config_path, endpoints_url, check_interval_s):
+        ...
```
Comment (rahul-tuli) [inline src/deepsparse/server/monitoring.py:61]: nice!

**Review by rahul-tuli — APPROVED** [review]

File: src/deepsparse/server/monitoring.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            continue
+
+        try:
+            _update_endpoints(endpoints_url, old_config, new_config)
+        except requests.RequestException:
+            yield None
+            _LOGGER.error("Requests to server failed, not updating.", exc_info=1)
+            continue
+
+        path = config_path + f".v{version}"
```
Comment (KSGulin) [inline src/deepsparse/server/monitoring.py:110]: Thoughts on saving previous versions of into a subdir next to the active config file, named "previous_configs" or something of that nature. Also, I think adding the date and time of modification to the config file name or as a markdown comment at the top of the config itself would be useful to users

Comment (corey-nm) [reply src/deepsparse/server/monitoring.py:110]: Like both of those, will do that!

Comment (corey-nm) [reply src/deepsparse/server/monitoring.py]: Added `<config path>.versions/` and a comment at the top of the file

**Review by KSGulin — DISMISSED** [review]
LGTM 🙂

File: src/deepsparse/server/config.py (hunk 2 above)
Comment (bfineran) [reply src/deepsparse/server/config.py:159]: +1 along with `old` and `new`

File: src/deepsparse/server/monitoring.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
```
Comment (bfineran) [inline src/deepsparse/server/monitoring.py:1]: let's rename this file to something more specific to the functionality given monitoring also has (more relevant) connotations to our logging functionality

**Review by bfineran — COMMENTED** [review]
lgtm, two minor comments. like the use of proc + requesting into the server

**Review by KSGulin — APPROVED** [review]

**Pull request merged**